File: ga/glm4_result_2024_6/result/message_queue_glm/49_message_queue.c
use std::io::{self, Read};
use std::process::{self, Child, Command};
use std::thread;
use std::time::Duration;

const MY_TYPE: i32 = 9527;

#[derive(Debug)]
struct MsgBuf {
    mtype: i32,
    mtext: String,
    number: i32,
}

fn main() {
    let mut input = String::new();
    println!("Please enter a string you want to send:");
    io::stdin().read_line(&mut input).expect("Failed to read line");
    let mtext: String = input.trim().to_string();

    println!("Please enter a number you want to send:");
    let mut number_input = String::new();
    io::stdin().read_line(&mut number_input).expect("Failed to read line");
    let number: i32 = number_input.trim().parse().expect("Failed to parse number");

    let msg = MsgBuf {
        mtype: MY_TYPE,
        mtext,
        number,
    };

    let mut child = Command::new("sh")
        .args(&["-c", "echo $MSG | tr -d '\n' | sed 's/[^0-9]//g' | wc -c"])
        .spawn()
        .expect("Failed to spawn child process");

    thread::sleep(Duration::from_secs(1));

    println!("Parent process sending message: {:?}", msg);
    let output = child
        .wait_with_output()
        .expect("Failed to wait on child process");

    println!("Child process output: {}", String::from_utf8_lossy(&output.stdout));
}

File: ga/gpt3_5_result_2024_6/result/message_queue_gpt/49_message_queue.rs
use std::mem;
use std::ffi::CString;
use std::os::raw::c_char;
use std::os::raw::c_int;
use std::ptr;
use std::process;

use libc::{msgget, msgsnd, msgrcv, IPC_PRIVATE, IPC_EXCL};

const MY_TYPE: i64 = 9527;

#[repr(C)]
struct MsgBuf {
    mtype: i64,
    mtext: [c_char; 100],
    number: c_int,
}

fn main() {
    let msgid = unsafe { msgget(IPC_PRIVATE, 0o666 | IPC_EXCL) };

    if msgid == -1 {
        eprintln!("Error in msgget");
        process::exit(-1);
    }

    let pid = unsafe { libc::fork() };

    if pid > 0 {
        // Parent process
        std::thread::sleep(std::time::Duration::from_secs(1));

        let mut buff: MsgBuf = unsafe { mem::zeroed() };

        buff.mtype = MY_TYPE;

        println!("Please enter a string you want to send:");
        let mut input_str = String::new();
        std::io::stdin().read_line(&mut input_str).expect("Failed to read line");
        let c_str = CString::new(input_str.trim()).expect("CString::new failed");
        let c_bytes = c_str.as_bytes_with_nul();
        unsafe {
            ptr::copy_nonoverlapping(c_bytes.as_ptr(), buff.mtext.as_mut_ptr() as *mut u8, c_bytes.len());
        }

        println!("Please enter a number you want to send:");
        let mut input_num = String::new();
        std::io::stdin().read_line(&mut input_num).expect("Failed to read line");
        let num: i32 = input_num.trim().parse().expect("Please enter a valid number");
        buff.number = num;

        unsafe {
            msgsnd(msgid, &buff as *const MsgBuf as *const std::ffi::c_void, std::mem::size_of::<MsgBuf>() - std::mem::size_of::<i64>(), 0);
        }

        unsafe {
            libc::waitpid(pid, std::ptr::null_mut(), 0);
        }
    } else if pid == 0 {
        // Child process
        let mut buff: MsgBuf = unsafe { mem::zeroed() };

        println!("Child process is waiting for msg:");
        unsafe {
            msgrcv(msgid, &mut buff as *mut MsgBuf as *mut std::ffi::c_void, std::mem::size_of::<MsgBuf>() - std::mem::size_of::<i64>(), MY_TYPE, 0);
        }

        let c_str = unsafe { std::ffi::CStr::from_ptr(buff.mtext.as_ptr()) };
        let str_slice = c_str.to_str().expect("Failed to convert C string to Rust string");

        println!("Child process read from msg: {}, {}", str_slice, buff.number);
    } else {
        eprintln!("Error in fork");
    }
}